### PR TITLE
Implement first few screens of Android 5.0-based UI.

### PR DESCRIPTION
**Pull request by sigmabeta**
For now, this is implemented as a separate option on the Android home screen - so while this UI is far from feature parity with the old one, they're installed side-by-side.

Looks kind of like this:

device-2015-05-06-201410
device-2015-05-06-201432

Will look nicer in the future (probably won't say "Dolphin New UI" in the title bar, for example.)

Screenshots are pulled from device storage, wherever Dolphin screenshots are saved by default. If you don't have any screenshot available for the game, a blank placeholder is used instead until you take one.

### CODE REVIEW

**Comment by sigmabeta** [comment]
Screenshots are pulled from device storage, wherever Dolphin screenshots are saved by default.

**Comment by Pringo** [comment]
I don't think using screenshots is a good approach for this. The majority of users won't have any and those that do still likely won't have them for every game leading to a bunch of empty placeholders for a UI.

**Comment by MaJoRoesch** [comment]
I agree. 99% of users will just have blanks, and the potential of whining to us about them. Banners are easily sourced, why can't we use those, and have screenshots as an optional setting for those that want the best appearance (as it does look really nice!)?

**Comment by rexcaptain501** [comment]
Totally agree with MaJo here, the more options the better.

**Comment by MaJoRoesch** [comment]
Btw, why those shades of blue in the title bar? Were you trying to use the colors from the logo? If so, it doesn't look right...

Here are the colors in case you want them
Etch border: #0057ab
Gradient top (light): #46d4ff
Gradient bottom (dark): #1792ff

Trying it out, I think the combination of Etch border for the top and gradient bottom for the title bar looks really good.

authenticdolphincolors

**Comment by lioncash** [comment]
There's tab-space mismatches in some places, and braces on the same line.

File: Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/model/GcGame.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+
+import java.io.File;
+
+public class GcGame implements Game {
+    private String mTitle;
+    private String mDescription;
+    private String mCountry;
+    private String mPath;
+    private String mGameId;
```
Comment (lioncash) [inline Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/model/GcGame.java]: These can likely be final (same with the class itself)

File: Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/model/WiiGame.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
+package org.dolphinemu.dolphinemu.model;
+
+
+public class WiiGame implements Game {
```
Comment (lioncash) [inline Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/model/WiiGame.java]: final class

**Comment by JosJuice** [comment]
* The apploader date only tells when the apploader was made, not when the game is released. Just displaying the apploader date as a date with no description like the UI does now is likely to make users think that it in fact is the release date. I suggest not displaying it at all.
* You should return an empty string instead of "Error" when data isn't available. It would in particular make the descriptions of Wii games look better.
* The description is too long to be displayed well in the grid view. (It's good on the detail screen, though.) Displaying the company instead is a possibility, or maybe the name could take up both lines? It looks a bit odd to me that not even a Wii game name (they have a strict character limit) like Super Mario Galaxy 2 fits right now.
* Only constructing a volume once would be a bit better than constructing it once for each piece of data (name, description, id...) but I don't know how easy it would be to do since I'm not familiar with the Android code.

**Comment by sigmabeta** [comment]
Thanks for the feedback everyone:

@Pringo The intention here is to create a UX in which a screenshot and save-state occur whenever the EmulationActivity stops, and the EmulationActivity defaults to loading the previous save-state. This flow much more closely resembles the expected behavior of regular Android apps (for example, screen rotations generally trigger a save->recreate screen->load flow, which is why some apps don't handle it correctly.) So the UI would silently pull from a cache of existing screenshots that is automatically generated. It just doesn't do it yet.

**Comment by sigmabeta** [comment]
@MaJoRoesch The problem with banners is that they're 96x32 in size. In the above screengrabs, the image size per card is roughly 430x342, and this was taken on a Nexus 9, which is probably one of the least pixel-dense screens available right now. On a modern phone, it would be in the ballpark of 860x700. If you look at the existing Android UI, it uses banners, and we run into this problem already - there just aren't enough pixels.

I would like to use banners though - in the second screengrab above, that circle with an X through it should contain the game's banner. I'd like to use the same circle motif in the actionbar for the EmulationActivity (which I believe would also get it to show up as the app's icon in Android's multitasking screen, but if not I'd like to make that happen as well.) I just need to implement a way of caching the banner to storage, so I don't have to call into native code to display the banner every time.

**Comment by sigmabeta** [comment]
@MaJoRoesch re: colors - I tried to mix together some mishmash of the Dolphin logo colors and the [Material Design spec colors](http://www.google.com/design/spec/style/color.html). I am not a graphic designer (I have some idea of how to make a usable app, but not necessarily a pretty one.) I vastly prefer the colors from your first suggestion; I'll definitely roll those in there.

Also, an idea that I am kicking around is to split the game list into a Gamecube list and a Wii list (hence the inheritance), divided into tabs, and having the Gamecube list use a different color scheme - have a look at colors.xml to see what I mean, and please suggest better colors for that too if you have them.

**Comment by sigmabeta** [comment]
@JosJuice The idea here was to show what an Android equivalent of the desktop's Properties dialog might *look* like, and I just wanted to throw some information on there regardless of how useful it was. That particular field is not terribly useful, so I'll probably remove it.

re: Error - the idea here was to catch your eye specifically :) the descriptions that returned "Error" were not doing so before your PR was merged the other day, and in fact I can't get the app to load at all unless I comment out the language-selection code you added. (My guess is that the language selection *would* work if I was only trying to load Wind Waker; i.e. these problems are probably related.) I'd like your help fixing this. But yes, showing "Error" is clearly not something I'm putting into a shipping product.

re: Titles not fitting - I'm experimenting with the Android 5.0 method of showing objects in a grid (RecyclerView, as opposed to GridView) which makes some changes to how things are scaled to fit, etc. In my UI editor, SMG2's title fits; that it does not at run-time is a bug that I'll fix.

re: Constructing volumes - I'm no JNI expert; any improvements you can make on that side are very welcome. However, my intention here is to read all the metadata about these files into an SQLite database on the Android side - so, all these native calls would be made only once. I'm not sure it's worth optimizing that, other than to make that initial import faster.

**Comment by JosJuice** [comment]
Wii games not having any descriptions is intended behavior. I wrote about why in the description of PR #2286. GameCube games should however be working just like before. Metroid Prime shows up as "Error" in your screenshot, but I think that's because it simply doesn't have a description, so it should be empty on all Dolphin versions and even real GameCubes.

As for the GetCurrentLanguage problem, all that I know about it is what I heard on IRC yesterday, which is that `SConfig::GetInstance().m_LocalCoreStartupParameter` might not be initialized correctly at the point which GetCurrentLanguage is called. It's unlikely to be related to the other problem. I might be able to take a look at it, but I can't test anything since I can't run the Android version and it works on both PC versions.

Caching the metadata once it's read (like the PC GUIs are doing) sounds great, but I still think the process is worth optimizing. The cache needs to be invalidated on Dolphin updates that change game metadata, so it can't always be used, especially for developers/testers that jump back and forth to different versions often. What happens right now if you try to read the name, description, company and banner from a GC game is that the volume will be opened, the filesystem table will be read from the disc, a representation of the entire filesystem will be created in memory, the filesystem will be scanned for opening.bnr, opening.bnr will be read from the disc... and then it discards everything and has to repeat it all three more times.

Maybe you could have a LoadVolumeMetadata function that reads all metadata at once and stores it to variables, roughly like LoadBanner did with m_names and m_is_wii_title before? The other functions could then read from the variables instead of the volume. It feels a little bit less clean since it relies on global state, but it might not possible to do much better unless you can pass an IVolume object through JNI or you make a function that returns all the metadata at once.

Also, Wii banners are 192x64 instead of 96x32, but I guess it doesn't make enough of a difference.

**Comment by sigmabeta** [comment]
Didn't know that about Wii banners.

I think the least messy approach is probably going to be to find a way to return all the relevant data in a single JNI call, but I'm not sure without looking into it how possible that is. We'll figure something out though.

Shame about Wii descriptions. All 10 of my Wii games, except for Metroid Other M, currently show descriptions in my few-weeks old PC build. You're saying we can't show those, even at the risk of having some strange games say weird stuff like "Save Data" in that field?

**Comment by JosJuice** [comment]
I chose to not show them because I thought it would be better that way. There aren't any technical barriers to re-implementing it.

**Comment by sigmabeta** [comment]
Cool. My position is that unless we're talking some huge proportion of games that just show garbage there, it's worth it. It's just a cosmetic thing anyway, and I'm willing to take "looks kind of cool most of the time, but kind of wacky sometimes", but maybe folks above my paygrade disagree.

Otherwise, I can just display a dialog without descriptions for Wii games - this was already kind of the plan.

Regarding company names, though - is there a Maker ID database somewhere that I could use to turn the Maker ID field into useful information?

**Comment by JosJuice** [comment]
http://www.gametdb.com/WiiCompanies

Is there an interest in having the IDs converted to company strings? If so, I could add it to the Wii volume code so that it becomes available in all GUIs.

**Comment by sigmabeta** [comment]
That would certainly make my work easier.

**Comment by BhaaLseN** [comment]
Why don't you create a JNI func that returns a Java class with all those properties set? It's not pretty on the C++ side, but should help with the multiple invocations on IVolume.

And I think its a decent idea to explore separate tabs/views/whatever for GC, Wii and potentially WADs, with different visual cues (GC being purple, Wii being...err...some other color) and different representations (Wii games don't have descriptions, so we can put something else there...if we find something worthwhile).

**Comment by sigmabeta** [comment]
The code style issues *should* be fixed. @lioncash , if you could merge it, I'd like to incorporate the above feedback (and further features) into new PRs.

**Comment by shuffle2** [comment]
fwiw I really like the idea of transparently using statesaves to "pause" the games...as long as it works correctly of course :)

File: Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/dialogs/GameDetailsDialog.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+import com.squareup.picasso.Picasso;
+
+import org.dolphinemu.dolphinemu.BuildConfig;
+import org.dolphinemu.dolphinemu.R;
+import org.dolphinemu.dolphinemu.emulation.EmulationActivity;
+import org.dolphinemu.dolphinemu.model.Game;
+
+import de.hdodenhof.circleimageview.CircleImageView;
+
+public class GameDetailsDialog extends DialogFragment {
```
Comment (lioncash) [inline Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/dialogs/GameDetailsDialog.java]: Brace on next line (rest of the file needs this too).

**Comment by lioncash** [comment]
You may want to put Javadoc comments on top of some of the class declarations. e.g. GameViewHolder explaining what its purpose is. It would make the codebase slightly easier to jump into for newcomers

**Comment by sigmabeta** [comment]
I will definitely add Javadocs for all the classes I've added, but I'm hesitant to do it with this PR because a lot of this functionality is going to change. For example, GameGridActivity will almost certainly have most of its functionality moved into a GameGridFragment.

**Comment by lioncash** [comment]
Oh ok, it's fine to hold off on Javadocs then.

**Comment by Pringo** [comment]
@sigmabeta a couple thoughts:

I think utilizing automatic save states is a really interesting idea but it does bring up a few potential problems that should be figured out first:
- Saves states are only valid for the one particular version of Dolphin in which they were generated. I can easily envision a scenario where a user relies on these saves states due to their convenience factor but then when they upgrade Dolphin to a newer version they end up losing all of their data. I think there should be some kind of preventative measure taken to hopefully stop this from occurring. 
- Considering the limited storage space on mobile devices how much room would all these screenshots and save states take up for a large game library? Would it be viable to keep these screenshots separate from the ones taken manually and resize and/or further compress them to reduce their storage footprint?

**Comment by sigmabeta** [comment]
@Pringo Those are very valid concerns.

* Re: storage, the idea is to just have the one automatic save state and screenshot per game. At ~50MB per state and <5MB per screenshot, I don't think this is unreasonable if the user has committed to a few 1.5-4.5GB games on their mobile device. It'll also be less of an issue on Android TV, where the user may have a permanently attached external hard drive. 
* Re: save state compatibility, that's a tough one. Short of someone coming up with native functionality to allow "do you want to try loading a potentially incompatible save state? yolo / no" to happen, I think the only thing we can do is educate the user and cross fingers. Potentially a [Snackbar](http://www.google.com/design/spec/components/snackbars-toasts.html#snackbars-toasts-specs) that shows the first time a state is loaded, and then every 2nd or 3rd time after that (with a "don't show again" button after the first time.) Text could say something like "Your save states will be deleted after every Dolphin update - don't forget to save in-game." and the "Don't show" setting could be tracked per-game. This solution is intrusive and a little bit annoying maybe, but I think easily outweighed by the convenience and improved UX.

**Comment by sigmabeta** [comment]
For the record: the solution we apply to PC - to display a message when the user tries to load an incompatible state directing them to the appropriate version number - is not feasible on Android.

When we have a proper CI setup for distributing to the Play Store, it will involve incrementing the APK's "versionCode" value with every published build. Both at the Play Store and the device level, a rule is enforced forbidding APK's with lower version codes from being installed. So rolling back, almost no matter what we do, will be a giant inconvenience for the user.

File: Source/Core/DolphinWX/MainAndroid.cpp [hunk 4]
```diff
@@ -287,12 +437,12 @@ JNIEXPORT jboolean JNICALL Java_org_dolphinemu_dolphinemu_NativeLibrary_Supports
 
 JNIEXPORT void JNICALL Java_org_dolphinemu_dolphinemu_NativeLibrary_SaveScreenShot(JNIEnv *env, jobject obj)
 {
-	Core::SaveScreenShot();
+Core::SaveScreenShot();
```
Comment (waddlesplash) [inline Source/Core/DolphinWX/MainAndroid.cpp]: Uh. What's up with this? This looks like a mistake. (Same below.)

Comment (sigmabeta) [inline Source/Core/DolphinWX/MainAndroid.cpp]: *sigh* I'm using Jetbrains' C++ IDE, which despite having a "1.0" version tag, mysteriously seems to have no concept whatsoever of what is going on with all these JNI methods. So the code style autoformatter seems to have barfed on it. I've already fixed this in the next PR I'm putting up.

Comment (waddlesplash) [inline Source/Core/DolphinWX/MainAndroid.cpp]: OK, good.

**Pull request merged**